Automattic/wp-calypso
# Track Events: Fix template part track events not firing

### Context

Fixes https://github.com/Automattic/wp-calypso/issues/62924.

If you go the Site Editor, select a template, and add or replace a template part (via the template part block), there are separate Tracks events that should fire for each of those actions: 

- wpcom_block_editor_template_part_choose_existing --> fires when adding a Template Part block and selecting "Choose"
- wpcom_block_editor_template_part_replace --> fires when you select a Template Part block, click replace, and pick a new part

Those events have stopped firing. The reason is that [Gutenberg 12.7.0](https://github.com/WordPress/gutenberg/releases/tag/v12.7.0) updated the UI and thus selectors that our tracking code relied on. 

In addition, that release also consolidated the UI for both actions above. Before, the UI and selectors were different. Now they use the exact same components, so you can no longer differentiate the two actions above based on selector. 

### Proposed Changes

1) I've consolidated the tracking related code for the two events from two files into a single file. Because these events now rely on the same selector with conditional logic, it was DRY-er to manage them together.

2) I've updated selectors for both events to reflect the new UI. Note that both events now rely on the same selector. 

3) I've added conditional logic that determines which of the two events will fire. A capture event handlers runs prior to any changes and records details of template part block in `replaceData`. Later, when we run the tracking code, if that value of replaceData is undefined, it means we're inserting a new template part; if it is defined, it means we're replacing a template part.

### Testing Instructions

Here's a video walk through of what everything should look like once fixed:

https://user-images.githubusercontent.com/21228350/180883682-64cd19d1-2d39-4de6-b1dd-339a05b97d20.mov


Note: the affected code here is part of the wpcom-block-editor. So when testing, you'll need to cd into the wp-calypso/apps/wpcom-block-editor directory and use the `yarn dev --sync` command. 

**Duplicate the issue.** You may want to duplicate the issue first. To see the issue you can do the following. 
1. Checkout the trunk branch of calypso (and the wpcom-editor-plugin). Be sure to run `yarn install` from the the calypso root and `yarn dev --sync` from the wpcom-block-editor directory. 
2. Open the Site Editor and select any template to edit.
3. Click the + sign to add a block and select the Template Part block.
4. At this point, open your dev tools, go to the Network tab, and clear all entries so you can monitor new network activity. Also in the search/filter box, type `t.gif`. This will filter network activity so you can see Tracks events.
5. In the Template Part block, click the "Choose" button and clic  an available template part to insert.
6. In the network tab of you dev tools, note that you do not see a Tracks event. **This is the bug.**
7. In the same template part block, click the "Replace" button and click on a new template part. 
8. In the network tab of you dev tools, note that you do not see a Tracks event. **This is the bug.**

**Confirm the fix.** Now checkout the fix/template-part-tracking branch, and repeat steps 1-8 above.
- In step 6, you should now see a Tracks event in the network tab of your dev tools. Click on the event, look at the Payload, and look for the _en property. You should see wpcom_block_editor_template_part_choose_existing. 
- In step 8, you now see the wpcom_block_editor_template_part_replace event. As an extra check, confirm that that the replaced_template_part_id property shows the old template part you replaced, and that the template_part_id property shows the new template part you've selected.
